src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -0,0 +1,183 @@
+        }
+    }
+
+    rd_kafka_conf_set_dr_msg_cb(conf, [](rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void * /* opaque */) {
+        static_cast<WriteBufferFromKafka *>(msg->_private)->onDrMsg(msg);
+    });
+
+    producer = RdKafkaPtr(rd_kafka_new(RD_KAFKA_PRODUCER, conf, errstr, sizeof(errstr)), rd_kafka_destroy);
+    if (!producer)
+    {
+        // librdkafka will take the ownership of `conf` if `rd_kafka_new` succeeds,
+        // but if it does not, we need to take care of cleaning it up by ourselves.
+        rd_kafka_conf_destroy(conf);
+        throw Exception("Failed to create kafka handle", klog::mapErrorCode(rd_kafka_last_error()));
+    }
+
+    auto topic = RdKafkaTopicPtr(rd_kafka_topic_new(producer.get(), kafka->topic().c_str(), nullptr), rd_kafka_topic_destroy);
+
+    wb = std::make_unique<WriteBufferFromKafka>(log, std::move(topic));
+
+    std::string data_format = kafka->dataFormat();
+    if (data_format.empty())
+        data_format = "JSONEachRow";
+
+    writer = FormatFactory::instance().getOutputFormat(data_format, *wb, header, context);
+    writer->setAutoFlush();
+
+    pollingThread = std::thread([this]() {
+        while (!isFinished)
+        {
+            auto n = rd_kafka_poll(producer.get(), 500 /* ms */);
+            LOG_TRACE(log, "polled {} events", n);
+        }
+    });
+}
+
+void KafkaSink::consume(Chunk chunk)
+{
+    if (!chunk.hasRows())
+        return;
+
+    auto blk = Block(getHeader());
+    blk.setColumns(chunk.getColumns());
+    writer->write(blk);
+}
+
+void KafkaSink::onFinish()
+{
+    isFinished = true;
+    if (pollingThread.joinable())
+        pollingThread.join();
+
+    // Make sure all outstanding requests are transmitted and handled.
+    if (auto err = rd_kafka_flush(producer.get(), -1 /* blocks */); err)
+        throw Exception(std::string("Failed to flush kafka producer: ") + rd_kafka_err2str(err), klog::mapErrorCode(err));
+
+    if (auto err = wb->deliveryError(); err != RD_KAFKA_RESP_ERR_NO_ERROR)
+        throw Exception(std::string("Failed to send messages: ") + rd_kafka_err2str(err), klog::mapErrorCode(err));
+
+    // if flush does not return an error, the delivery report queue should be empty
+    if (!wb->fullyAcked())
+        throw Exception(
+            fmt::format("Not all messsages are acked, expected={} actual={}", wb->outstandings(), wb->acked()),
+            ErrorCodes::MISSING_ACKNOWLEDGEMENT);
+}
+
+KafkaSink::~KafkaSink()
+{
+    onFinish();
+}
+
+void KafkaSink::checkpoint(CheckpointContextPtr context)
+{
+    do
+    {
+        if (auto err = wb->deliveryError(); err != RD_KAFKA_RESP_ERR_NO_ERROR)
+            throw Exception(std::string("Failed to send messages: ") + rd_kafka_err2str(err), klog::mapErrorCode(err));
+
+        if (wb->fullyAcked())
+            break;
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    } while (true);
```

Comment:
We may need check the shutdown / finished since it will be dead loop if there is something wrong in external Kafka, and we can't tear down.